### PR TITLE
Use RNGCryptoServiceProvider instead of System.Random

### PR DESCRIPTION
**Pull request by scottbommarito**
https://devdiv.visualstudio.com/DevDiv/_workitems?id=244741&_a=edit

This was the only use of System.Random in the code base.

I don't know how to verify that this works as intended in the Search Client but I verified the code that I changed generates random numbers within the correct range.

This is not entirely fair, but I think the difference in fairness is probably very minimal.

What I mean is that when you generate only discrete bytes of randomness rather than discrete chunks of the size you want, you get uneven randomness. For example, if you were looking for a range of 0-2 but were generating 0-3 and used modulus when the generated number was too big, 0 would turn up 50% of the time and 1 and 2 would only show up 25% of the time each. However, I think because we are generating 8 bit chunks instead of 2 bits, the effect is much more minimal.

Also I didn't want to loop generating numbers that were out of range because this method is much slower than System.Random.

Let me know if you think there's a better solution to this or we can find a way to get rid of the randomness altogether.

### CODE REVIEW

**Comment by maartenba** [comment]
The randomness is part of our load balancing here. Not sure if we have to kill of `System.Random` here?

**Comment by scottbommarito** [comment]
@maartenba Unfortunately the SDL task says "use only approved random number generators", not "use only approved random number generators when doing security-critical work"...so I think it's either do this or ask for an exception...

**Comment by maartenba** [comment]
Ship it 😀

**Comment by scottbommarito** [comment]
(forgot to tag more people)
@skofman1 @shishirx34 @ryuyu

File: src/NuGet.Services.Search.Client/Client/RetryingHttpClientWrapper.cs [hunk 1]
```diff
@@ -216,17 +216,26 @@ private static bool ShouldTryOther(HttpResponseMessage response)
...
-                var randomNumber = _random.Next(0, totalWeight);
+
+                int bytesRequired = (int)Math.Ceiling(Math.Log(totalWeight) / Math.Log(2));
+                var randomBytes = new byte[bytesRequired];
+                int randomNumber;
+                
+                _random.GetBytes(randomBytes);
+                randomNumber = Math.Abs(BitConverter.ToInt32(randomBytes, 0)) % totalWeight;
+                // This is not entirely fair, but it's close enough.
+                // See https://msdn.microsoft.com/en-us/library/wb9c8c67(v=vs.110).aspx for more info.
```
Comment (skofman1) [inline src/NuGet.Services.Search.Client/Client/RetryingHttpClientWrapper.cs]: please separate this logic into a method, and add UTs for this method, making sure that the number you return is in the correct range

**Comment by skofman1** [comment]
:clock1:

**Comment by blowdart** [comment]
The SDL only applies for crypto-operations :) 

"This requirement applies to all Microsoft products and services _that use random numbers for security purposes_"

If it's not for security, then Random() is just fine.

**Comment by scottbommarito** [comment]
@blowdart lol thanks

**Comment by blowdart** [comment]
Don't let this stop you going through the SDL though :) This deserves a :cookie:

**Pull request closed without merging**